keps/sig-autoscaling/2021-scale-from-zero/README.md

```diff
@@ -0,0 +1,839 @@
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: `HPAScaleToZero`
+    - Components depending on the feature gate: `kube-apiserver`, `kubelet`
+  - [x] Other
+    - Describe the mechanism:
+
+      When HPAScaleToZero feature gate is enabled HPA supports scaling to zero pods based on object or external metrics. HPA remains active as long as at least one metric value available.
+
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+
+      No
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+      No
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+  Any change of default behavior may be surprising to users or break existing
+  automations, so be extremely careful here.
+
+  HPA creation/update with `minReplicas: 0` is no longer rejected.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+  Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+  Describe the consequences on existing workloads (e.g., if this is a runtime
+  feature, can it break the existing applications?).
+
+  Yes. To downgrade the cluster to version that does not support scale-to-zero feature:
+
+  1. Make sure there are no hpa objects with minReplicas=0 and maxReplicas=0. Here is a oneliner to update it to 1:
+
+      `$ kubectl get hpa --all-namespaces  --no-headers=true | awk  '{if($6==0) printf "kubectl patch hpa/%s --namespace=%s -p \"{\\\"spec\\\":{\\\"minReplicas\\\":1,\\\"maxReplicas\\\":1}}\"\n", $2, $1 }' | sh`
+  2. Disable `HPAScaleToZero` feature gate
```

Comment:
On kube-apiserver and kubelet both, does it matter? Out of curiousity, why does kubelet need the feature gate?

What happens if you do this and there ARE HPA objects with min/max of 0/0?

Reply:
> Out of curiousity, why does kubelet need the feature gate?

It seems not to be needed, removed.